File: testbed_php/test/associations/OptionalOneToNTest.php
<?php

class OptionalOneToNTest extends UnitTestCase
{


  public function test_CreateStudentWithoutMentor()
  {
    $s = new StudentD();
    $this->assertEqual(null,$s->getMentor());
  }
  
  public function test_ConstructorWithoutNStudents()
  {
  	$this->expectException(new Exception("Unable to create MentorD, must have 3 students"));
    $s = new StudentD();
    $s2 = new StudentD();    
    new MentorD(array($s,$s2));
  }
  

  public function test_setStudents_makeSureAlwaysN()
  {
    $s = new StudentD();
    $s2 = new StudentD();
    $s3 = new StudentD();

    $m = new MentorD(array($s,$s2,$s3));
    $s4 = new StudentD();
    $s5 = new StudentD();
    $s6 = new StudentD();
    
    $this->assertEqual(true,$m->setStudents(array($s, $s2, $s3)));
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual($m,$s3->getMentor());

    $this->assertEqual(true,$m->setStudents(array($s4, $s5, $s6)));
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual(null,$s->getMentor());
    $this->assertEqual(null,$s2->getMentor());
    $this->assertEqual(null,$s3->getMentor());
    $this->assertEqual($m,$s4->getMentor());
    $this->assertEqual($m,$s5->getMentor());
    $this->assertEqual($m,$s6->getMentor());
  }

  public function test_setStudents_watchOutForDuplicateEntries()
  {
  	$this->expectException(new Exception("Unable to create MentorD, must have 3 students"));
    $s = new StudentD();
    $s2 = new StudentD();
    new MentorD(array($s, $s2, $s));
  }
  
  

  public function test_setStudents_replaceExisitingmakeSureAlwaysN()
  {
    $s = new StudentD();
    $s2 = new StudentD();
    $s3 = new StudentD();

    $m = new MentorD(array($s, $s2, $s3));
    $s4 = new StudentD();
    
    $this->assertEqual(false,$m->setStudents(array($s, $s2)));
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual($m,$s3->getMentor());
    $this->assertEqual(null,$s4->getMentor());    
    
    $this->assertEqual(false,$m->setStudents(array($s, $s2, $s3, $s4)));
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual(null,$s4->getMentor());

  }


  public function test_setStudents_empty()
  {
    $s = new StudentD();
    $s2 = new StudentD();
    $s3 = new StudentD();
    $m = new MentorD(array($s,$s2,$s3));

    
    $this->assertEqual(true,$m->setStudents(array($s, $s2, $s3)));
    $this->assertEqual(3,$m->numberOfStudents());

    $this->assertEqual(false,$m->setStudents(array()));
    $this->assertEqual(3,$m->numberOfStudents());
  }
  
  public function test_checkForExistingMentor()
  {
		$this->expectException(new Exception("Unable to create MentorD, must have 3 students"));
    $s1 = new StudentD();
    $s2 = new StudentD();
    $s3 = new StudentD();

    $m = new MentorD(array($s1,$s2,$s3));

    $s4 = new StudentD();
    
    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual($m, $s1->getMentor());
    $this->assertEqual($m, $s2->getMentor());
    $this->assertEqual($m, $s3->getMentor());    
    $this->assertEqual(null, $s4->getMentor());    

    new MentorD(array($s1,$s2,$s4));
  }   
  

  public function test_deletStudentDeletesTheMentor()
  {
    $s1 = new StudentD();
    $s2 = new StudentD();
    $s3 = new StudentD();

    $m = new MentorD(array($s1,$s2,$s3));
    $studentP = new ProgramD();
    $mentorP = new ProgramD();
    
    $s2->setProgram($studentP);
    $m->setProgram($mentorP);

    $s1->delete();
    
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual(null,$m->getProgram());
    $this->assertEqual(null,$mentorP->getMentor());
    $this->assertEqual($studentP,$s2->getProgram());
    $this->assertEqual(null,$s1->getMentor());
  }  

  
  
}

File: testbed_ruby/test/associations/optional_one_to_n_test.rb
require 'test/unit'
require './src-gen-umple/student_d'
require './src-gen-umple/mentor_d'
require './src-gen-umple/program_d'

module CruiseAssociations
class OptionalOneToNTest < Test::Unit::TestCase

  def test_CreateStudentWithoutMentor
    s = StudentD.new
    assert_equal(nil,s.get_mentor)
  end
  
  def test_ConstructorWithoutNStudents
    assert_raise RuntimeError do #"Unable to create MentorD, must have 3 students"
      s = StudentD.new
      s2 = StudentD.new    
      MentorD.new([s,s2])
    end
  end
  
  def test_set_students_makeSureAlwaysN
    s = StudentD.new
    s2 = StudentD.new
    s3 = StudentD.new
  
    m = MentorD.new([s,s2,s3])
    s4 = StudentD.new
    s5 = StudentD.new
    s6 = StudentD.new
    
    assert_equal(true,m.set_students([s, s2, s3]))
    assert_equal(3,m.number_of_students)
    assert_equal(m,s.get_mentor)
    assert_equal(m,s2.get_mentor)
    assert_equal(m,s3.get_mentor)
  
    assert_equal(true,m.set_students([s4, s5, s6]))
    assert_equal(3,m.number_of_students)
    assert_equal(nil,s.get_mentor)
    assert_equal(nil,s2.get_mentor)
    assert_equal(nil,s3.get_mentor)
    assert_equal(m,s4.get_mentor)
    assert_equal(m,s5.get_mentor)
    assert_equal(m,s6.get_mentor)
  end
  
  def test_set_students_watchOutForDuplicateEntries
    assert_raise RuntimeError do #"Unable to create MentorD, must have 3 students"
      s = StudentD.new
      s2 = StudentD.new
      MentorD.new([s, s2, s])
    end
  end
  
  def test_set_students_replaceExisitingmakeSureAlwaysN
    s = StudentD.new
    s2 = StudentD.new
    s3 = StudentD.new
  
    m = MentorD.new([s, s2, s3])
    s4 = StudentD.new
    
    assert_equal(false,m.set_students([s,s2]))
    assert_equal(3,m.number_of_students)
    assert_equal(m,s.get_mentor)
    assert_equal(m,s2.get_mentor)
    assert_equal(m,s3.get_mentor)
    assert_equal(nil,s4.get_mentor)    
    
    assert_equal(false,m.set_students([s, s2, s3, s4]))
    assert_equal(3,m.number_of_students)
    assert_equal(nil,s4.get_mentor)
  end
  
  def test_set_students_empty
    s = StudentD.new
    s2 = StudentD.new
    s3 = StudentD.new
    m = MentorD.new([s,s2,s3])
  
    
    assert_equal(true,m.set_students([s, s2, s3]))
    assert_equal(3,m.number_of_students)
  
    assert_equal(false,m.set_students([]))
    assert_equal(3,m.number_of_students)
  end
  
  def test_checkForExistingMentor
    s1 = StudentD.new
    s2 = StudentD.new
    s3 = StudentD.new

    m = MentorD.new([s1,s2,s3])

    s4 = StudentD.new
  
    assert_equal(3,m.number_of_students)
    assert_equal(m, s1.get_mentor)
    assert_equal(m, s2.get_mentor)
    assert_equal(m, s3.get_mentor)    
    assert_equal(nil, s4.get_mentor)    

    assert_raise RuntimeError do #"Unable to create MentorD, must have 3 students"
      MentorD.new([s1,s2,s4])
    end
  end   
  
  def test_deletStudentDeletesTheMentor
    s1 = StudentD.new
    s2 = StudentD.new
    s3 = StudentD.new
  
    m = MentorD.new([s1,s2,s3])
    studentP = ProgramD.new
    mentorP = ProgramD.new
    
    s2.set_program(studentP)
    m.set_program(mentorP)
  
    s1.delete
    
    assert_equal(0,m.number_of_students)
    assert_equal(nil,m.get_program)
    assert_equal(nil,mentorP.get_mentor)
    assert_equal(studentP,s2.get_program)
    assert_equal(nil,s1.get_mentor)
  end  

end
end